modules/kinesis/src/main/scala/com/snowplowanalytics/snowplow/enrich/kinesis/Sink.scala

```diff
@@ -14,164 +14,208 @@ package com.snowplowanalytics.snowplow.enrich.kinesis
-    val retryPolicy = capDelay[F](config.backoffPolicy.maxBackoff, fullJitter[F](config.backoffPolicy.minBackoff))
-      .join(limitRetries(config.backoffPolicy.maxRetries))
-    val partitionKey = data.attributes.toList match { // there can be only one attribute : the partition key
-      case head :: Nil => head._2
-      case _ => UUID.randomUUID().toString
-    }
-    val res = for {
-      byteBuffer <- Async[F].delay(ByteBuffer.wrap(data.data))
-      cb <- producer.putData(config.streamName, partitionKey, byteBuffer)
-      cbRes <- registerCallback(cb)
-      _ <- ContextShift[F].shift
-    } yield cbRes
-
-    res
-      .retryingOnFailuresAndAllErrors(
-        wasSuccessful = _.isSuccessful,
+    val withoutRetry = blocker.blockOn(Sync[F].delay(putRecords(kinesis, config.streamName, records)))
+
+    val withRetry =
+      withoutRetry.retryingOnAllErrors(
         policy = retryPolicy,
-        onFailure = (result, retryDetails) =>
-          Logger[F]
-            .warn(
-              List(
-                s"Writing to shard ${result.getShardId()} failed.",
-                s"${result.getAttempts().size()} KPL attempts (${retryDetails.retriesSoFar} retries from cats-retry).",
-                s"Errors : [${getErrorMessages(result)}]"
-              ).mkString(" ")
-            ),
         onError = (exception, retryDetails) =>
           Logger[F]
-            .error(exception)(s"Writing to Kinesis errored (${retryDetails.retriesSoFar} retries from cats-retry)")
+            .error(exception)(
+              s"Writing ${records.size} records to ${config.streamName} errored (${retryDetails.retriesSoFar} retries from cats-retry)"
+            )
       )
-      .void
+
+    for {
+      _ <- retryStatus match {
+             case RetryStatus.NoRetriesYet =>
+               Logger[F].debug(s"Writing ${records.size} records to ${config.streamName}")
+             case retry =>
+               Logger[F].debug(
+                 s"Waiting for ${retry.cumulativeDelay} and retrying to write ${records.size} records to ${config.streamName}"
+               ) *>
+                 Timer[F].sleep(retry.cumulativeDelay)
+           }
+      putRecordsResult <- withRetry
+      failuresRetried <- if (putRecordsResult.getFailedRecordCount != 0) {
+                           val failurePairs = records.zip(putRecordsResult.getRecords.asScala).filter(_._2.getErrorMessage != null)
+                           val (failedRecords, failedResults) = failurePairs.unzip
+                           val logging = logErrorsSummary(getErrorsSummary(failedResults))
+                           val maybeRetrying =
+                             for {
+                               nextRetry <- retryPolicy.decideNextRetry(retryStatus)
+                               maybeRetried <- nextRetry match {
+                                                 case PolicyDecision.DelayAndRetry(delay) =>
+                                                   writeToKinesis(
```

Comment:
A more elegant solution would be explicitly tail recursive.  It's hurting my brain to work out if this is susceptible to stack overflow exception -- I can't tell.

Luckily the number of retries is small, so we're not going to get stack overflow.  For that reason I'm OK if you don't change anything for now.

Reply:
Good plan!

Reply:
https://github.com/snowplow/enrich/issues/644